cpp/src/ice2slice/Main.cpp

```diff
@@ -0,0 +1,303 @@
+    for(vector<string>::const_iterator i = optargs.begin(); i != optargs.end(); ++i)
+    {
+        cppArgs.push_back("-D" + *i);
+    }
+
+    optargs = opts.argVec("U");
+    for(vector<string>::const_iterator i = optargs.begin(); i != optargs.end(); ++i)
+    {
+        cppArgs.push_back("-U" + *i);
+    }
+
+    vector<string> includePaths = opts.argVec("I");
+    for(vector<string>::const_iterator i = includePaths.begin(); i != includePaths.end(); ++i)
+    {
+        cppArgs.push_back("-I" + Preprocessor::normalizeIncludePath(*i));
+    }
+
+    bool preprocess = opts.isSet("E");
+
+    string output = opts.optArg("output-dir");
+
+    bool debug = opts.isSet("debug");
+
+    bool ice = opts.isSet("ice");
+
+    bool underscore = opts.isSet("underscore");
+
+    if(args.empty())
+    {
+        consoleErr << argv[0] << ": error: no input file" << endl;
+        if(!validate)
+        {
+            usage(argv[0]);
+        }
+        return EXIT_FAILURE;
+    }
+
+    if(validate)
+    {
+        return EXIT_SUCCESS;
+    }
+
+    int status = EXIT_SUCCESS;
+
+    IceUtil::CtrlCHandler ctrlCHandler;
+    ctrlCHandler.setCallback(interruptedCallback);
+
+    ostringstream os;
+
+    // Create a copy of args without the duplicates.
+    vector<string> sources;
+    for(vector<string>::const_iterator i = args.begin(); i != args.end(); ++i)
+    {
+        vector<string>::iterator p = find(sources.begin(), sources.end(), *i);
+        if(p == sources.end())
+        {
+            sources.push_back(*i);
+        }
+    }
+
+    map<string, vector<string> > moduleInfo;
+
+    for(vector<string>::const_iterator i = sources.begin(); i != sources.end();)
+    {
+        PreprocessorPtr icecpp = Preprocessor::create(argv[0], *i, cppArgs);
+        FILE* cppHandle = icecpp->preprocess(true, "-D__ICE2SLICE__");
+
+        if(cppHandle == 0)
+        {
+            return EXIT_FAILURE;
+        }
+
+        if(preprocess)
+        {
+            char buf[4096];
+            while(fgets(buf, static_cast<int>(sizeof(buf)), cppHandle) != ICE_NULLPTR)
+            {
+                if(fputs(buf, stdout) == EOF)
+                {
+                    return EXIT_FAILURE;
+                }
+            }
+            if(!icecpp->close())
+            {
+                return EXIT_FAILURE;
+            }
+        }
+        else
+        {
+            UnitPtr p = Unit::createUnit(false, false, ice, underscore);
+            int parseStatus = p->parse(*i, cppHandle, debug);
+
+            if(!icecpp->close())
+            {
+                p->destroy();
+                return EXIT_FAILURE;
+            }
+
+            if(parseStatus == EXIT_FAILURE)
+            {
+                status = EXIT_FAILURE;
+            }
+            else
+            {
+                DefinitionContextPtr dc = p->findDefinitionContext(p->topLevelFile());
+                assert(dc);
+                const string prefix = "js:module:";
```

Comment:
Why this `js:module` processing?

Reply:
Need to remove, took slice2js as a template to start this project.